### PR TITLE
Pepc boost impl 2

### PR DESCRIPTION
**Pull request by bharath-123**
## 📝 Summary

1. We allow the builder to send the blocks to a new relayer API `/relay/v1/builder/rob_blocks`. We ideally want to merge `/relay/v1/builder/rob_blocks` API with `/relay/v1/builder/blocks` since `/relay/v1/builder/blocks` is a special case of a ROB bid without additional TOB txs. But we will keep them separate for development purposes.
2. We add a block assembler endpoint(flashbots_blockAssembler), which takes in the TOB txs list and a ROB payload bid. It assembles the txs and returns a final execution payload. 

The block assembler endpoint checks for duplicate txs in the TOB and ROB tx list. It drops the duplicate txs in the ROB list. (We have not yet supported it for now to test things, but it will be an easy change)

The ROB payload withdrawals list is used for the final payload.

We add a new field to the `miner.BuilderPayloadArgs` called `AssemblerTxLists`; this struct holds two new fields, `TobTxs` and `RobTxs`. These lists are used in the `generateWork` method to build a block with txs from the TOB and ROB. The block is returned to the relayer via the `flashbots_blockAssembler` endpoint

## 📚 References

Design Doc: https://docs.google.com/document/d/1DLse86yGLId2idQUIJbaiGxY_RxMxz3LhtdinbA3q4Q/edit

### CODE REVIEW

File: builder/builder.go [hunk 1]
```diff
@@ -47,7 +47,9 @@ type ValidatorData struct {
 
 type IRelay interface {
 	SubmitBlock(msg *bellatrixapi.SubmitBlockRequest, vd ValidatorData) error
+	SubmitRobBlock(msg *bellatrixapi.SubmitBlockRequest, vd ValidatorData) error
```
Comment (michaelneuder) [inline builder/builder.go]: dont need to support bellatrix

File: builder/builder.go [hunk 2]
```diff
@@ -202,9 +204,10 @@ func (b *Builder) Stop() error {
 
 func (b *Builder) onSealedBlock(block *types.Block, blockValue *big.Int, ordersClosedAt, sealedAt time.Time,
 	commitedBundles, allBundles []types.SimulatedBundle, usedSbundles []types.UsedSBundle,
-	proposerPubkey phase0.BLSPubKey, vd ValidatorData, attrs *types.BuilderPayloadAttributes) error {
+	proposerPubkey phase0.BLSPubKey, vd ValidatorData, attrs *types.BuilderPayloadAttributes, isRobBlock bool) error {
 	if b.eth.Config().IsShanghai(block.Time()) {
-		if err := b.submitCapellaBlock(block, blockValue, ordersClosedAt, sealedAt, commitedBundles, allBundles, usedSbundles, proposerPubkey, vd, attrs); err != nil {
+		log.Info("DEBUG: Submitting capella block\n")
```
Comment (michaelneuder) [inline builder/builder.go:210]: can just log.Debug

Comment (michaelneuder) [reply builder/builder.go:210]: here and below

File: builder/builder.go [hunk 3]
```diff
@@ -325,10 +328,20 @@ func (b *Builder) submitCapellaBlock(block *types.Block, blockValue *big.Int, or
...
 	} else {
 		go b.ds.ConsumeBuiltBlock(block, blockValue, ordersClosedAt, sealedAt, commitedBundles, allBundles, usedSbundles, &blockBidMsg)
-		err = b.relay.SubmitBlockCapella(&blockSubmitReq, vd)
-		if err != nil {
-			log.Error("could not submit capella block", "err", err, "#commitedBundles", len(commitedBundles))
-			return err
+		if isRobBlock {
+			log.Info("DEBUG: submitting rob capella block")
+			err = b.relay.SubmitRobBlockCapella(&blockSubmitReq, vd)
+			if err != nil {
```
Comment (michaelneuder) [inline builder/builder.go:335]: id pull this out of the conditional b/c it repeats in the else. just call the SubmitBlock functions in the conditional and assign to an outer err variable

File: builder/eth_service.go [hunk 4]
```diff
@@ -54,6 +54,7 @@ func NewEthereumService(eth *eth.Ethereum) *EthereumService {
 func (s *EthereumService) BuildBlock(attrs *types.BuilderPayloadAttributes, sealedBlockCallback miner.BlockHookFn) error {
 	// Send a request to generate a full block in the background.
 	// The result can be obtained via the returned channel.
+	log.Info("DEBUG: In build block!!")
```
Comment (michaelneuder) [inline builder/eth_service.go:57]: id remove this comment and these types of logs. they look like print statements

File: builder/local_relay.go [hunk 5]
```diff
@@ -66,6 +66,16 @@ type LocalRelay struct {
 	fd            ForkData
 }
 
+func (r *LocalRelay) SubmitRobBlock(msg *bellatrixapi.SubmitBlockRequest, vd ValidatorData) error {
```
Comment (michaelneuder) [inline builder/local_relay.go]: dont need bellatrix support

File: builder/local_relay.go [hunk 6]
```diff
@@ -66,6 +66,16 @@ type LocalRelay struct {
...
+	//TODO implement me
+	panic("implement me")
+}
+
+func (r *LocalRelay) SubmitRobBlockCapella(msg *capellaapi.SubmitBlockRequest, vd ValidatorData) error {
+	//TODO implement me
+	panic("implement me")
```
Comment (michaelneuder) [inline builder/local_relay.go:71]: i wouldnt put a panic here. just return an error

File: builder/relay.go [hunk 7]
```diff
@@ -71,7 +71,7 @@ func (r *RemoteRelay) updateValidatorsMap(currentSlot uint64, retries int) error
 	r.validatorSyncOngoing = true
 	r.validatorsLock.Unlock()
 
-	log.Info("requesting ", "currentSlot", currentSlot)
+	//log.Info("DEBUG: requesting from relayer", "currentSlot", currentSlot)
```
Comment (michaelneuder) [inline builder/relay.go:74]: remove

File: builder/relay.go [hunk 8]
```diff
@@ -99,10 +99,12 @@ func (r *RemoteRelay) GetValidatorForSlot(nextSlot uint64) (ValidatorData, error
 	// next slot is expected to be the actual chain's next slot, not something requested by the user!
 	// if not sanitized it will force resync of validator data and possibly is a DoS vector
 
+	//log.Info("DEBUG: Getting validator for slot", "slot", nextSlot)
```
Comment (michaelneuder) [inline builder/relay.go:102]: remove commented out logs

File: builder/relay.go [hunk 9]
```diff
@@ -155,6 +157,24 @@ func (r *RemoteRelay) SubmitBlock(msg *bellatrix.SubmitBlockRequest, _ Validator
 	return nil
 }
 
+func (r *RemoteRelay) SubmitRobBlock(msg *bellatrix.SubmitBlockRequest, _ ValidatorData) error {
```
Comment (michaelneuder) [inline builder/relay.go]: dont need

File: builder/relay_aggregator.go [hunk 10]
```diff
@@ -18,6 +18,16 @@ type RemoteRelayAggregator struct {
 	registrationsCache     map[ValidatorData][]IRelay
 }
 
+func (r *RemoteRelayAggregator) SubmitRobBlock(msg *bellatrix.SubmitBlockRequest, vd ValidatorData) error {
```
Comment (michaelneuder) [inline builder/relay_aggregator.go]: dont need

File: builder/relay_aggregator.go [hunk 11]
```diff
@@ -18,6 +18,16 @@ type RemoteRelayAggregator struct {
...
+	//TODO implement me
+	panic("implement me")
+}
+
+func (r *RemoteRelayAggregator) SubmitRobBlockCapella(msg *capella.SubmitBlockRequest, vd ValidatorData) error {
+	//TODO implement me
+	panic("implement me")
```
Comment (michaelneuder) [inline builder/relay_aggregator.go]: no panic

File: core/types/transaction.go [hunk 12]
```diff
@@ -591,6 +599,15 @@ func NewSBundleWithMinerFee(sbundle *SimSBundle, _ *big.Int) (*TxWithMinerFee, e
 	}, nil
 }
 
+// Used only to assemble txs. During tx assembly, we need to have txs in the sequence specified by assembly request
```
Comment (michaelneuder) [inline core/types/transaction.go:602]: remove commented out code

File: core/types/withdrawal.go [hunk 13]
```diff
@@ -45,6 +45,10 @@ type withdrawalMarshaling struct {
 // Withdrawals implements DerivableList for withdrawals.
 type Withdrawals []*Withdrawal
 
+func (s Withdrawals) Append(w Withdrawal) Withdrawals {
```
Comment (michaelneuder) [inline core/types/withdrawal.go:48]: dont think you need this ? append should work on the type natively

File: eth/block-validation/api.go [hunk 14]
```diff
@@ -228,6 +233,172 @@ func (r *BuilderBlockValidationRequestV2) UnmarshalJSON(data []byte) error {
...
+	if err != nil {
+		return err
+	}
+	b.RegisteredGasLimit = intermediateJson.RegisteredGasLimit
+	b.RobPayload = intermediateJson.RobPayload
+
+	return nil
+}
+
+// bchain: copied this here to avoid circular dependency
```
Comment (michaelneuder) [inline eth/block-validation/api.go:264]: hmm can you refactor to a shared common package instead ? copying feels a bit bad

File: eth/block-validation/api.go [hunk 15]
```diff
@@ -228,6 +233,172 @@ func (r *BuilderBlockValidationRequestV2) UnmarshalJSON(data []byte) error {
...
+			return nil, errors.New("duplicate tx")
+		}
+		seenTxMap[tx.Hash()] = struct{}{}
+	}
+	log.Info("Done Checking for duplicate txs!")
+
+	withdrawals := make(types.Withdrawals, len(params.RobPayload.ExecutionPayload.Withdrawals))
+	log.Info("DEBUG: Building the final set of withdrawals")
+	log.Info("DEBUG: Withdrawals are ", "withdrawals", params.RobPayload.ExecutionPayload.Withdrawals)
+	for i, withdrawal := range params.RobPayload.ExecutionPayload.Withdrawals {
```
Comment (michaelneuder) [inline eth/block-validation/api.go:350]: why is this necessary ? withdrawals should all be the same regardless. why do we have to copy like this

File: eth/block-validation/api_test.go [hunk 16]
```diff
@@ -720,6 +1030,19 @@ func TestValidateBuilderSubmissionV2_CoinbasePaymentDefault(t *testing.T) {
 	require.ErrorContains(t, api.ValidateBuilderSubmissionV2(req), "payment")
 }
 
+//func TestBlockAssembler(t *testing.T) {
```
Comment (michaelneuder) [inline eth/block-validation/api_test.go:1033]: remove

File: miner/worker.go [hunk 17]
```diff
@@ -1095,6 +1102,139 @@ func (w *worker) commitBundle(env *environment, txs types.Transactions, interrup
...
+
+		tx := assemblerTxs.RobTxs.Index(i)
+
+		// Error may be ignored here. The error has already been checked
+		// during transaction acceptance is the transaction pool.
+		from, _ := types.Sender(env.signer, tx)
+		logs, err := w.commitTransaction(env, tx)
+		log.Info("DEBUG: ROB tx commit failed with err", "err", err)
+		switch {
+		case errors.Is(err, core.ErrGasLimitReached):
```
Comment (michaelneuder) [inline miner/worker.go:1190]: super repetitive code, can you refactor this into a shared function

**Review by michaelneuder — COMMENTED** [review]
First round of comments: 

high level feedback
- need some sort of description of going on / more detailed design diagrams. its pretty tough to review this b/c there are so many files and its hard to see how they fit together.
- no need to support bellatrix stuff
- remove commented out stuff and debug statements for reviewability
- can you split this into smaller PRs? where each PR does a specific thing or defines a specific type. if you chain them together it will be so much easier to follow the threads.
- high level diagram of the flow of the ToB and RoB txns through the block validation flow and the pruning stage could be hugely benefitial

File: builder/builder.go (hunk 1 above)
Comment (bharath-123) [reply builder/builder.go]: true

File: miner/worker.go (hunk 17 above)
Comment (bharath-123) [reply miner/worker.go:1190]: yeah yeah will do. i didn't focus too much on code quality for now tbh. can do these things

File: eth/block-validation/api_test.go (hunk 16 above)
Comment (bharath-123) [reply eth/block-validation/api_test.go:1033]: yes

File: eth/block-validation/api.go (hunk 15 above)
Comment (bharath-123) [reply eth/block-validation/api.go:350]: its some golang type thing dude. will see if i can make it look cleaner

File: eth/block-validation/api.go (hunk 14 above)
Comment (bharath-123) [reply eth/block-validation/api.go:264]: yeah yeah will do it.

File: builder/builder.go (hunk 2 above)
Comment (bharath-123) [reply builder/builder.go:210]: will remove the logs for review

Comment (bharath-123) [reply builder/builder.go:210]: will remove log statements for review

File: builder/builder.go (hunk 3 above)
Comment (bharath-123) [reply builder/builder.go:335]: I am planning to merge this with SubmitBlock itself for the review. I initially intended to separate them for simplicity. But for review, I ll merge both the APIs. We can always make a 2nd API in the future

File: builder/local_relay.go (hunk 6 above)
Comment (bharath-123) [reply builder/local_relay.go:71]: lot of places have panics in the code, I guess its a standard

File: eth/block-validation/api.go (hunk 15 above)
Comment (bharath-123) [reply eth/block-validation/api.go:350]: conversions b/w types. there are some opaque types in the code which makes it a bit weird.
